napari/components/dims.py

```diff
@@ -116,100 +63,83 @@ class Dims:
         ``displayed`` dimensions.
     """
 
+    # fields
     ndim: int = 2
-    ndisplay: Property[int, None, only_2D_3D] = 2
+    ndisplay: Literal[2, 3] = 2
     last_used: int = 0
+    range: Tuple[Tuple[float, float, float], ...] = ()
+    current_step: Tuple[int, ...] = ()
+    order: Tuple[int, ...] = ()
+    axis_labels: Tuple[str, ...] = ()
+
+    # private vars
+    _scroll_progress: int = 0
+
+    # validators
+    @validator('axis_labels', pre=True)
+    def _string_to_list(v):
+        if isinstance(v, str):
+            return list(v)
+        return v
+
+    @root_validator
+    def _check_dims(cls, values):
```

Comment:
Could you add a docstring here? This function is doing a lot of work and as someone not familiar with pydantic and root_validator, I just have no idea what it's doing. Comments for each of the steps would be nice too, e.g.:

```python
# check that there is a range for each dimension.
# if there are too few ranges, prepend (0, 2, 1) for each missing dimension.
# if there are too many, throw out the leading ranges.
# this is done because ... (🤷)
```

Reply:
I realise looking below that this was in some sense existing code... But it's never too late to improve our docs! ;)

Reply:
comments and doc string added